File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.java
class Pair {
    public char letter;
    public int count;
    
    Pair(char letter, int count) {
        this.letter = letter;
        this.count = count;
    }
}
class Solution {
    public String frequencySort(String s) {
        int[] freq = new int[256];
        for(int i = 0 ; i < s.length(); i++) {
            int ascii = s.charAt(i);
            freq[ascii]++;
        }
        
        Queue<Pair> pq = new PriorityQueue<>((Pair a, Pair b) -> b.count - a.count);
        
        for(int i = 0; i < 256; i++) {
            if(freq[i] > 0) {
                pq.offer(new Pair( (char) i, freq[i]));
            }
        }
                     
        StringBuilder result = new StringBuilder();
        while(!pq.isEmpty()) {
            Pair top = pq.poll();
            for (int i = 0 ; i < top.count; i++) {
                result.append(top.letter);
            }
        }
        
        return result.toString();
    }
}

File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.cpp
class Solution {
public:
    string frequencySort(string s) {
        int freq[256] = {};
        for(auto c: s) 
            freq[c]++;
        sort(s.begin(), s.end(), [&](char a , char b){
            return freq[a] != freq[b] ? freq[a] > freq[b] : a > b;
        });
        return s;
    }
};